File: LeetCode/5. Longest Palindromic Substring.rb
def longest_palindrome(s)
    return "" if s.eql? ""
    ans = s[0]
    s.each_char.with_index do |char, index|
        p1 = expandAroundCenter(s, index, index)
        p2 = expandAroundCenter(s, index, index+1)
        ans = p1 if p1.length > ans.length
        ans = p2 if p2.length > ans.length
    end
    ans
end

def expandAroundCenter(s, left, right)
    while left >= 0 and right < s.length and s[left] == s[right]
        left -= 1
        right += 1
    end
    s[left+1, right-left-1]
end

File: LeetCode/5. Longest Palindromic Substring.cpp
class Solution {
public:
    string expandAroundCenter(string s, int left, int right) {
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        return s.substr(left + 1, right - (left + 1));
    }

    string longestPalindrome(string s) {
        if (s == "") return "";
        if (s.length() == 1) return s;

        string ans = "";

        for (int i = 0; i < s.length() - 1; i++) {
            string p = expandAroundCenter(s, i, i);
            string q = expandAroundCenter(s, i, i + 1);

            if (ans.length() < q.length() || ans.length() < p.length()) {
                ans = (p.length() > q.length()) ? p : q;
            }
        }
        return ans;
    }
};
